### PR TITLE
Honor `config.output_router_logits` in the MoE VLM wrappers

### PR DESCRIPTION
**Pull request by qgallouedec**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48885) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48885)

Four `ForConditionalGeneration` wrappers read the flag straight off `kwargs` instead of resolving it against the config, so `config.output_router_logits = True` does nothing and only a forward argument turns the auxiliary loss on. Every text-only MoE honors both.

`Qwen3MoeForCausalLM` is the reference:

https://github.com/huggingface/transformers/blob/0a959de1d2dd0c981f1f732dbd0fc31192bbfa66/src/transformers/models/qwen3_moe/modeling_qwen3_moe.py#L640-L642
https://github.com/huggingface/transformers/blob/0a959de1d2dd0c981f1f732dbd0fc31192bbfa66/src/transformers/models/qwen3_vl_moe/modular_qwen3_vl_moe.py#L422
https://github.com/huggingface/transformers/blob/0a959de1d2dd0c981f1f732dbd0fc31192bbfa66/src/transformers/models/glm4v_moe/modular_glm4v_moe.py#L384
https://github.com/huggingface/transformers/blob/0a959de1d2dd0c981f1f732dbd0fc31192bbfa66/src/transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py#L2068
https://github.com/huggingface/transformers/blob/0a959de1d2dd0c981f1f732dbd0fc31192bbfa66/src/transformers/models/qwen4_exp/modeling_qwen4_exp.py#L2525

Qwen3.5-MoE and Qwen4-Exp carry both patterns in the same file, the text model resolving against the config and the wrapper ignoring it: 
https://github.com/huggingface/transformers/blob/0a959de1d2dd0c981f1f732dbd0fc31192bbfa66/src/transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py#L1897-L1899
vs https://github.com/huggingface/transformers/blob/0a959de1d2dd0c981f1f732dbd0fc31192bbfa66/src/transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py#L2068

After, all three agree with the reference, and an explicit `output_router_logits=False` still wins over a config `True`.

## Changes

Resolve the argument against the text config before the backbone call, the way `Qwen3MoeForCausalLM.forward` does, and pass it down explicitly. Two modular files cover all four, since `Qwen3_5MoeForConditionalGeneration(Qwen3VLMoeForConditionalGeneration)` and `Qwen4ExpForConditionalGeneration(Qwen3_5MoeForConditionalGeneration)`.

The two configs also need the field:

- `Qwen3VLMoeTextConfig` had `output_router_logits` deleted in #41250, correctly at the time, since nothing read it: https://github.com/huggingface/transformers/blob/0a959de1d2dd0c981f1f732dbd0fc31192bbfa66/src/transformers/models/qwen3_vl_moe/modular_qwen3_vl_moe.py#L119 It is read now, so the deletion is dropped.
- `Glm4vMoeTextConfig` never had it. It inherits from `Glm4MoeConfig`, and GLM4-MoE balances its experts without an auxiliary loss. GLM4V-MoE does compute one, so the field is added next to the `router_aux_loss_coef` the config already carries: https://github.com/huggingface/transformers/blob/0a959de1d2dd0c981f1f732dbd0fc31192bbfa66/src/transformers/models/glm4v_moe/modular_glm4v_moe.py#L109

One test per modular source, asserting the config default is honored and that the explicit argument wins over it.

## Why it matters

Downstream code that decides "can this model produce the auxiliary loss?" from the config gets the wrong answer in both directions. This is what TRL's `router_aux_loss_coef` was keyed on (huggingface/trl#7222): the coefficient was silently ignored on Qwen3-VL-MoE and GLM4V-MoE.

## Not covered here

This aligns how the flag is read, not what every MoE does with it. Four models still expose `aux_loss` in their output type and never populate it, so `outputs.aux_loss` is `None` even with the flag on:

| | `output_router_logits` argument | returns `router_logits` | computes `aux_loss` |
|---|---|---|---|
| AFMoE | yes | yes | no |
| Zaya | yes | yes | no |
| Cohere2-MoE | no | yes | no |
| HY-V3 | no | yes | no |

`HYV3ForCausalLM` is explicit about it: `aux_loss=None,  # Not used in this model`. Either the loss should be computed from the router logits these models already return, or the field should be dropped from their output types. That is a per-architecture question, so it belongs in its own PR.

### CODE REVIEW

File: tests/models/glm4v_moe/test_modeling_glm4v_moe.py [hunk 1]
```diff
@@ -205,6 +205,19 @@ def test_model_is_small(self):
...
+        model = Glm4vMoeForConditionalGeneration(config).to(torch_device).eval()
+        input_ids = inputs_dict["input_ids"]
+
+        assert model(input_ids).aux_loss is None
+        assert model(input_ids, output_router_logits=True).aux_loss is not None
+
+        model.config.get_text_config().output_router_logits = True
+        assert model(input_ids).aux_loss is not None
+        assert model(input_ids, output_router_logits=False).aux_loss is None
+
```
Comment (zucchini-nlp) [inline tests/models/glm4v_moe/test_modeling_glm4v_moe.py]: i thought we had a test in common mixin, but seeing that the PR never got merged due to merge-queue checkd (https://github.com/huggingface/transformers/pull/46642)

Might be worth to move into mixin which enforces for all models

Comment (zucchini-nlp) [reply tests/models/glm4v_moe/test_modeling_glm4v_moe.py]: ah also the linked PR added aux-loss in gemma4 family, i suppose it is still missing so let's add (if it is still needed)

File: tests/models/glm4v_moe/test_modeling_glm4v_moe.py [hunk 2]
```diff
@@ -205,6 +205,19 @@ def test_model_is_small(self):
...
         self.config_tester.run_common_tests()
 
+    def test_output_router_logits_from_config(self):
+        """`config.output_router_logits` turns the auxiliary loss on, and an explicit argument wins over it."""
+        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+        model = Glm4vMoeForConditionalGeneration(config).to(torch_device).eval()
+        input_ids = inputs_dict["input_ids"]
+
+        assert model(input_ids).aux_loss is None
+        assert model(input_ids, output_router_logits=True).aux_loss is not None
```
Comment (Cyrilvallez) [inline tests/models/glm4v_moe/test_modeling_glm4v_moe.py]: As @zucchini-nlp said, would be better as a general test instead!

**Review by Cyrilvallez — APPROVED** [review]
LGTM, let's just make the test general instead of model specific.
We can either merge both PR (the one linked by @zucchini-nlp) or merge separately, no strong opinion. 
One important thing to consider is that the general test from the linked PR relies on the config attr to be present to trigger the test, which is exactly what both PR fix -> so would not detect faulty configs. Maybe we should find a better way to detect MoEs to have a more robust test, maybe if `"router_logits"`is in `_can_record_outputs` is a more robust way

**Comment by qgallouedec** [comment]
Moved to `ModelTesterMixin` as `test_output_router_logits_from_config`. 

It runs on every model whose config has `output_router_logits` and whose head declares `router_logits` in its output type, checks that the config flag turns the logits on (and the aux loss where the model computes one) and that an explicit argument wins over it.

Gemma4: it has no `output_router_logits` at all, so there is nothing for this PR to honor. Adding the aux loss there is a modeling change, separate PR.

File: tests/test_modeling_common.py [hunk 3]
```diff
@@ -5870,6 +5871,47 @@ def new_reset(self, token):
...
+            self.skipTest("This model has no `output_router_logits` in its config.")
+
+        for model_class in self.all_model_classes:
+            with self.subTest(model_class.__name__):
+                return_type = model_class.forward.__annotations__.get("return")
+                output_fields = set().union(
+                    *(getattr(t, "__dataclass_fields__", {}).keys() for t in (get_args(return_type) or (return_type,)))
+                )
+                if "router_logits" not in output_fields:
+                    self.skipTest(f"{model_class.__name__} does not declare router_logits in its output type.")
```
Comment (zucchini-nlp) [inline tests/test_modeling_common.py:5889]: unrelated to this pr -  if config has `output_router_logits`, it should prob have the field in outputs of BaseModels

Or if this is to guard a base model vs generative model, then we can filter `all_model_names` by `get_values(MODEL_MAPPING_NAMES)`

**Review by zucchini-nlp — APPROVED** [review]
A few tests failing, can merge after fixing them

**Comment by qgallouedec** [comment]
Fixed the 4 failures in 5afac8bb: JetMoe ignored `config.output_router_logits` (same bug as the VLMs, now resolved against the config), and musicgen's composite config needs `get_text_config(decoder=True)` in the common test. The Qwen4-Exp TP failure is a crash in `ple_embedding` on the text-only model, which this PR doesn't touch.

run-slow: glm4v_moe, jetmoe, minimax_m3_vl, qwen3_5_moe, qwen3_vl_moe, qwen4_exp
